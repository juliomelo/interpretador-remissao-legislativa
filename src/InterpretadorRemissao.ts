import InterpretadorReferencia, { IReferenciaEncontrada, TipoReferencia } from './InterpretadorReferencia';
import IRemissao, { IReferenciaItem } from './IRemissao';
import ITipoNorma from './ITipoNorma';

export interface IInterpretacaoRemissao {
    idx: number;
    remissao: IRemissao;
}

export interface IReferencia {
    [tipo: string]: IReferenciaEncontrada;
}

export default class InterpretadorRemissao {
    private hashNormas: Map<string, ITipoNorma>;
    private regexp: RegExp;
    private interpretadorDispositivo = new InterpretadorReferencia();

    constructor(normas: ITipoNorma[]) {
        normas = [
            {
                ambito: 'Federal',
                tipo: 'Constituição Federal',
                sigla: 'CF',
                semNumero: true
            },
            {
                ambito: 'Federal',
                tipo: 'Constituição da República Federativa do Brasil',
                sigla: 'CRFB',
                semNumero: true
            },
            {
                ambito: 'Estadual',
                tipo: 'Constituição do Estado',
                sigla: 'CE',
                semNumero: true
            },
            {
                ambito: 'Federal',
                tipo: 'Constituição Estadual',
                sigla: 'CE',
                semNumero: true
            },
            {
                ambito: 'Municipal',
                tipo: 'Lei Orgânica',
                sigla: 'LO',
                semNumero: true
            },
            ...normas];
        this.hashNormas = new Map();

        for (const norma of normas) {
            this.hashNormas.set(norma.tipo.trim().toLocaleLowerCase(), norma);
            this.hashNormas.set(norma.sigla.trim().toLocaleLowerCase(), norma);
        }

        const regexpInicio = '(^|\\s|[.:;,!?()[\\]{}`´"\'\u2018\u2019\u201C\u201D])';
        const regexpNormas = '(' +
            normas.map(norma => norma.tipo.trim().replace(/\s+/g, '\\s+'))
                .concat(normas.map(norma => norma.sigla.trim()))
                .sort((a, b) => b.length - a.length)
                .join('|') + ')';
        const regexpNumero = '(?:(?:n[º.]?|n[uú]mero|n[uú]m\\.)\\s*)?(\\d+(?:\\.\\d{3})+)';
        const regexpDiaMes = '\\s*de\\s*\\d+\\s*de(?:\\s*\\d+\\s*|\\s+\\S+\\s+)de\\s*';
        const regexpAno = `(?:(?:[/-:]|,?${regexpDiaMes}|,?\\s*?de\\s*?|\\s+)(\\d+)(?!\\s*de\\s*\\S+\\s*de\\s*\\d+))`;
        const regexpFinal = '($|\\s|[.:;,!?()[\\]{}`´"\'\u2018\u2019\u201C\u201D])';

        // tslint:disable-next-line: max-line-length
        this.regexp = new RegExp(`${regexpInicio}${regexpNormas}(?:\\s*${regexpNumero}${regexpAno}?)?${regexpFinal}`, 'ig');
    }

    public interpretar(entrada: string): IInterpretacaoRemissao[] {
        const remissoes = this.interpretarRessissaoNormas(entrada);

        remissoes.forEach(remissao => this.interpretarRemissaoDispositivos(entrada, remissao));

        return remissoes;
    }

    private interpretarRessissaoNormas(entrada: string): IInterpretacaoRemissao[] {
        const resultado: IInterpretacaoRemissao[] = [];

        for (let m = this.regexp.exec(entrada); m; m = this.regexp.exec(entrada)) {
            const tipo = this.hashNormas.get(m[2].trim().toLowerCase())!;
            const idx = m.index + m[1].length;
            const texto = m[0].substr(m[1].length, m[0].length - m[1].length - m[5].length);

            if (m[3]) {
                resultado.push({
                    idx,
                    remissao: {
                        texto,
                        identificador: {
                            tipo,
                            numero: parseInt(m[3].replace(/\./g, ''), 10),
                            ano: parseInt(m[4], 10)
                        }
                    }
                });
            } else if (tipo.semNumero) {
                resultado.push({
                    idx,
                    remissao: {
                        texto,
                        identificador: {
                            tipo
                        }
                    }
                });
            }
        }

        return resultado;
    }

    private interpretarRemissaoDispositivos(entrada: string, remissao: IInterpretacaoRemissao): void {
        let idx = remissao.idx - 1;

        if (idx < 0) {
            return;
        }

        const referencia: IReferencia = {};
        let inicio = idx;

        for (const item of this.interpretadorDispositivo.interpretarReverso(entrada, idx)) {
            if (item.tipo in referencia) {
                console.debug(`Referência ${item.tipo} repetida em ${item.idx}.`);
                this.incorporarReferencia(remissao.remissao, referencia, entrada, inicio, idx - inicio + 1);
                idx = inicio - 1;
            }

            referencia[item.tipo] = item;
            inicio = Math.min(inicio, item.idx);
        }

        if (inicio !== idx) {
            this.incorporarReferencia(remissao.remissao, referencia, entrada, inicio, idx - inicio + 1);
        }
    }

    private incorporarReferencia(remissao: IRemissao,
                                 referencia: IReferencia,
                                 entrada: string,
                                 inicio: number,
                                 tamanho: number) {
        if (!remissao.referencia) {
            remissao.referencia = [];
        }

        remissao.referencia.unshift({
            artigo: this.extrairNumero(entrada, referencia[TipoReferencia.ARTIGO]),
            paragrafo: this.extrairNumero(entrada, referencia[TipoReferencia.PARAGRAFO], false),
            // tslint:disable-next-line: max-line-length
            inciso: this.extrairNumero(entrada, referencia[TipoReferencia.INCISO], TipoReferencia.ALINEA in referencia),
            // tslint:disable-next-line: max-line-length
            alinea: this.extrairNumero(entrada, referencia[TipoReferencia.ALINEA], TipoReferencia.ITEM in referencia),
            item: this.extrairNumero(entrada, referencia[TipoReferencia.ITEM], false),
            texto: entrada.substr(inicio, tamanho)
        } as IReferenciaItem);
    }

    private extrairNumero(entrada: string,
                          referencia: IReferenciaEncontrada,
                          obrigatorio: boolean = true): string | undefined {
        if (!referencia) {
            if (obrigatorio) {
                throw new Error(`Referência incompleta.`);
            }

            return undefined;
        }

        const espaco = /\s|,/;
        let idx = referencia.idx + referencia.tamanho;

        while (idx < entrada.length && espaco.test(entrada.charAt(idx))) {
            idx++;
        }

        let ultimoIdx = idx + 1;

        while (ultimoIdx < entrada.length && !espaco.test(entrada.charAt(ultimoIdx))) {
            ultimoIdx++;
        }

        if (ultimoIdx >= entrada.length) {
            if (obrigatorio) {
                throw new Error(`Referência incompleta.`);
            }

            return undefined;
        } else {
            return entrada.substr(idx, ultimoIdx - idx).replace(/[\u2018\u2019\u201C\u201D'"]/g, '');
        }
    }
}
