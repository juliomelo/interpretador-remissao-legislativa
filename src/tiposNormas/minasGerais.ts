import ITipoNorma from '../ITipoNorma';

export default [
    {
        ambito: 'Minas Gerais',
        tipo: 'Ato das Disposições Constitucionais Transitórias',
        sigla: 'ADT'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Constituição',
        sigla: 'CON'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Decisão Conjunta',
        sigla: 'DCJ'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Decisão da Mesa',
        sigla: 'DCM'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Decisão da Presidência',
        sigla: 'DPR'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Decisão da Primeira-Secretaria',
        sigla: 'DEP'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Decisão Normativa da Presidência',
        sigla: 'DNP'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Decreto',
        sigla: 'DEC'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Decreto com numeração especial',
        sigla: 'DNE'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Decreto sem Número',
        sigla: 'DSN'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Decreto-Lei',
        sigla: 'DEL'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Deliberação',
        sigla: 'DLB'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Emenda à Constituição',
        sigla: 'EMC'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Instrução Normativa da Diretoria-Geral',
        sigla: 'IDG'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Lei',
        sigla: 'LEI'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Lei Adicional',
        sigla: 'LEA'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Lei Complementar',
        sigla: 'LCP'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Lei Constitucional',
        sigla: 'LCO'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Lei Delegada',
        sigla: 'LDL'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Ordem de Serviço',
        sigla: 'OSV'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Ordem de Serviço da Presidência, da Primeira-Secretaria e da Diretoria',
        sigla: 'OSP'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Ordem de Serviço da Presidência e da Primeira-Secretaria',
        sigla: 'OPP'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Ordem de Serviço da Primeira-Secretaria e da Diretoria-Geral',
        sigla: 'OPD'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Portaria',
        sigla: 'PRT'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Portaria da Diretoria-Geral',
        sigla: 'PDG'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Portaria da Presidência',
        sigla: 'PPR'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Portaria da Presidência, da Primeira-Secretaria e da Diretoria-Geral',
        sigla: 'PPD'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Portaria da Presidência e da Diretoria-Geral',
        sigla: 'PPG'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Portaria da Presidência e da Primeira-Secretaria',
        sigla: 'PPS'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Portaria da Primeira-Secretaria e da Diretoria-Geral',
        sigla: 'PSD'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Portaria sem Número',
        sigla: 'PSN'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Resolução',
        sigla: 'RAL'
    },
    {
        ambito: 'Minas Gerais',
        tipo: 'Resolução Sem Número',
        sigla: 'RSN'
    }
] as ITipoNorma[];
