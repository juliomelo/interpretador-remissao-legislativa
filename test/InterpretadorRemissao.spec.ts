import { readFileSync } from 'fs';
import InterpretadorRemissao from '../src/InterpretadorRemissao';
import minasGerais from '../src/tiposNormas/minasGerais';

describe('InterpretadorRemissao', () => {
    const normas = [...minasGerais, {
        ambito: 'Federal',
        tipo: 'Lei Federal',
        sigla: 'LF'
    }];
    const interpretador = new InterpretadorRemissao(normas);
    const interpretadorSegmentado = new InterpretadorRemissao(normas, { segmentarDispositivo: true });

    function testar(entrada: string) {
        expect(interpretador.interpretar(entrada)).toMatchSnapshot('Interpretador Padrão');
        expect(interpretadorSegmentado.interpretar(entrada)).toMatchSnapshot('Interpretador Segmentado');
    }

    it('Deve interpretar lei', () => {
        // tslint:disable-next-line: max-line-length
        testar('O § 6º do art. 225 da Lei nº 6.763, de 26 de dezembro de 1975, passa a vigorar com a seguinte redação:');
    });

    it('Não deve confundir referência de artigos entre diferentes períodos.', () => {
        // tslint:disable-next-line: max-line-length
        const texto = 'Testando o art. 225. Veja a Lei nº 6.763, de 26 de dezembro de 1975.';
        const resultado = interpretador.interpretar(texto);

        expect(resultado[0].remissao.referencia).toBe(undefined);
    });

    it('Deve interpretar remissões do Decreto com Numeração Especial 471 de 23/09/2019', () => {
        testar(readFileSync('test/minas_gerais-decreto_com_numeracao_especial_471_2019.txt').toString());
    });
});
