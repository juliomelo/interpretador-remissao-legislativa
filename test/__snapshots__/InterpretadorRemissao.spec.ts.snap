// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InterpretadorRemissao Deve interpretar lei 1`] = `
Array [
  Object {
    "idx": 22,
    "remissao": Object {
      "identificador": Object {
        "ano": 1975,
        "numero": 6763,
        "tipo": Object {
          "ambito": "Minas Gerais",
          "sigla": "LEI",
          "tipo": "Lei",
        },
      },
      "referencia": Array [
        Object {
          "alinea": undefined,
          "artigo": "225",
          "inciso": undefined,
          "item": undefined,
          "paragrafo": "6º",
          "texto": "§ 6º do art. 225 da ",
        },
      ],
      "texto": "Lei nº 6.763, de 26 de dezembro de 1975",
    },
  },
]
`;

exports[`InterpretadorRemissao Deve interpretar remissões do Decreto com Numeração Especial 471 de 23/09/2019 1`] = `
Array [
  Object {
    "idx": 81,
    "remissao": Object {
      "identificador": Object {
        "ano": 2006,
        "numero": 11428,
        "tipo": Object {
          "ambito": "Federal",
          "sigla": "LF",
          "tipo": "Lei Federal",
        },
      },
      "referencia": Array [
        Object {
          "alinea": "b",
          "artigo": "3º",
          "inciso": "VII",
          "item": undefined,
          "paragrafo": "3º",
          "texto": "alínea “b” do inciso VII do art. 3º e no ",
        },
        Object {
          "alinea": undefined,
          "artigo": "14",
          "inciso": undefined,
          "item": undefined,
          "paragrafo": "3º",
          "texto": "§ 3º do art. 14, ambos da ",
        },
      ],
      "texto": "Lei Federal nº 11.428, de 22 de dezembro de 2006",
    },
  },
]
`;
